## Author: Suraj Bhattarai     
## Ecological Genomics:   

### Overall Description of notebook      

This notebook is to store all the codings and assignments that we do in Ecological Genomics class.


### Date started: 2018-01-24  
### Date end:   (year-month-day)    

### Philosophy   
Science should be reproducible and one of the best ways to achieve this is by logging research activities in a notebook. Because science/biology has increasingly become computational, it is easier to document computational projects in an electronic form, which can be shared online through Github.    

### Helpful features of the notebook     

**It is absolutely critical for your future self and others to follow your work.**     

* The notebook is set up with a series of internal links from the table of contents.    
* All notebooks should have a table of contents which has the "Page", date, and title (information that allows the reader to understand your work).     
* Also, one of the perks of keeping all activities in a single document is that you can **search and find elements quickly**.     
* Lastly, you can share specific entries because of the three "#" automatically creates a link when the notebook renders on github.      


<a rel="license" href="http://creativecommons.org/licenses/by/4.0/"><img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by/4.0/88x31.png" /></a><br />This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution 4.0 International License</a>.  


### Table of contents for 60 entries (Format is *Page: Date(with year-month-day). Title*)        
* [Page 1: 2018-01-24](#id-section1). Intro to Github, RMarkdown, and UNIX command-line
* [Page 2: 2018-01-29](#id-section2). Working with RNAseq data of Beetles
* [Page 3:](#id-section3).
* [Page 4:](#id-section4).
* [Page 5:](#id-section5).
* [Page 6:](#id-section6).
* [Page 7:](#id-section7).
* [Page 8:](#id-section8).
* [Page 9:](#id-section9).
* [Page 10:](#id-section10).
* [Page 11:](#id-section11).
* [Page 12:](#id-section12).
* [Page 13:](#id-section13).
* [Page 14:](#id-section14).
* [Page 15:](#id-section15).
* [Page 16:](#id-section16).
* [Page 17:](#id-section17).
* [Page 18:](#id-section18).
* [Page 19:](#id-section19).
* [Page 20:](#id-section20).
* [Page 21:](#id-section21).
* [Page 22:](#id-section22).
* [Page 23:](#id-section23).
* [Page 24:](#id-section24).
* [Page 25:](#id-section25).
* [Page 26:](#id-section26).
* [Page 27:](#id-section27).
* [Page 28:](#id-section28).
* [Page 29:](#id-section29).
* [Page 30:](#id-section30).
* [Page 31:](#id-section31).
* [Page 32:](#id-section32).
* [Page 33:](#id-section33).
* [Page 34:](#id-section34).
* [Page 35:](#id-section35).
* [Page 36:](#id-section36).
* [Page 37:](#id-section37).
* [Page 38:](#id-section38).
* [Page 39:](#id-section39).
* [Page 40:](#id-section40).
* [Page 41:](#id-section41).
* [Page 42:](#id-section42).
* [Page 43:](#id-section43).
* [Page 44:](#id-section44).
* [Page 45:](#id-section45).
* [Page 46:](#id-section46).
* [Page 47:](#id-section47).
* [Page 48:](#id-section48).
* [Page 49:](#id-section49).
* [Page 50:](#id-section50).
* [Page 51:](#id-section51).
* [Page 52:](#id-section52).
* [Page 53:](#id-section53).
* [Page 54:](#id-section54).
* [Page 55:](#id-section55).
* [Page 56:](#id-section56).
* [Page 57:](#id-section57).
* [Page 58:](#id-section58).
* [Page 59:](#id-section59).
* [Page 60:](#id-section60).

------
<div id='id-section1'/>
### Page 1: 2018-01-24. Notes on using Github, Rmarkdown, and the UNIX command-line

To log into the pbio381 server, i use ssh. Like this:


Open PuTTY program and enter following commands: 
```
Using username "sbhatta4".
sbhatta4@pbio381.uvm.edu's password:
Last login: Wed Jan 24 11:11:22 2018 from ip0af520ce.int.uvm.edu
[sbhatta4@pbio381 ~]$
```
To see the full directory, use pwd command

```
[sbhatta4@pbio381 ~]$ pwd
/users/s/b/sbhatta4
[sbhatta4@pbio381 ~]$

```

To make a directory, use mkdir command

To view all the directories, use ll command

To enter into a directory, use cd 'directory name' command

cd /data takes you into the shared space where you might find some data

```
[sbhatta4@pbio381 mydata]$ cd /data
[sbhatta4@pbio381 data]$ ll
total 12
drwxr-xr-x.  5 root     root       73 Jan 31  2017 archive
drwxr-xr-x.  2 srkeller users      49 Jan 23 13:20 external_data
drwxrwxr-x.  2 root     pb381adm   40 Nov 30  2015 packages
drwxrwxr-x. 48 root     pb381adm 4096 Mar 28  2017 popgen
drwxrwxr-x.  4 root     pb381adm   42 Jan 23 16:51 project_data
drwxrwxr-x.  2 root     pb381adm 4096 Aug  8 16:53 scripts
drwxr-xr-x. 21 root     root     4096 Feb  1  2017 users


```
To look what's in those files write cd 'file name'/ For e.g to see what's inside project_data folder, write cd project_data/ and write ll 

```
[sbhatta4@pbio381 data]$ cd project_data
[sbhatta4@pbio381 project_data]$ ll
total 4
drwxrwxr-x. 4 root pb381adm   35 Jan 23 16:52 beetles
drwxr-xr-x. 8 root pb381adm 4096 Jan 23 13:20 ssw


```
There are two folders beetles and ssw. Similarly, to see what's inside beetles write cd beetles/ and then write ll
```
drwxr-xr-x. 6 root pb381adm 86 Jan 23 16:52 archive
drwxr-xr-x. 2 root pb381adm 34 Jan 23 16:51 metadata
[sbhatta4@pbio381 beetles]$


```

Now when we opened metadata, then we find cols_data.txt. this is a text file, so it must conatain some information about the beetles. But we don't want to edit these contents in the shared space. So we need to first copy this file into our personal space. So we can now use cp commnad to copy it to our home directory. 

cp cols_data.txt ~/mydata/

```

-rwxr-xr-x. 1 root pb381adm 1474 Jan 23 16:50 cols_data.txt
[sbhatta4@pbio381 metadata]$ cp cols_data.txt ~/mydata/
[sbhatta4@pbio381 metadata]$


```

Now to see whether it has been copied to our personal folder or now, we need to open our home directory and check into  mydata by using ll 

```
[sbhatta4@pbio381 metadata]$ cd ~/mydata/
[sbhatta4@pbio381 mydata]$ ll
total 4
-rwxr-xr-x. 1 sbhatta4 users 1474 Jan 24 14:16 cols_data.txt


```
Now we want to see what's inside this file. We can use 'head' command to see the first 10 lines and 'tail' command to see the last 10 lines. 

```
sample_ID       population      devstage        sex
IT_AD4_F1_      IT      AD4     F
IT_AD4_F2_      IT      AD4     F
IT_AD4_F3_      IT      AD4     F
IT_AD4_M1_      IT      AD4     M
IT_AD4_M2_      IT      AD4     M
IT_AD4_M3_      IT      AD4     M
IT_ADC_F13      IT      ADC     F
IT_ADC_F14      IT      ADC     F
IT_ADC_F15      IT      ADC     F
[sbhatta4@pbio381 mydata]$ tail cols_data.txt
WA_PD1_F3_      WA      PD1     F
WA_PD1_M1_      WA      PD1     M
WA_PD1_M2_      WA      PD1     M
WA_PD1_M3_      WA      PD1     M
WA_PP1_F1_      WA      PP1     F
WA_PP1_F2_      WA      PP1     F
WA_PP1_F3_      WA      PP1     F
WA_PP1_M1_      WA      PP1     M
WA_PP1_M2_      WA      PP1     M
WA_PP1_M3_      WA      PP1     M


```
If we want to search specific thing in the file, we can use search command grep. For e.g. if we want to see only the samples from IT, then we should use grep 'IT' cols_data.txt


```
[sbhatta4@pbio381 mydata]$ grep 'IT' cols_data.txt
IT_AD4_F1_      IT      AD4     F
IT_AD4_F2_      IT      AD4     F
IT_AD4_F3_      IT      AD4     F
IT_AD4_M1_      IT      AD4     M
IT_AD4_M2_      IT      AD4     M
IT_AD4_M3_      IT      AD4     M
IT_ADC_F13      IT      ADC     F
IT_ADC_F14      IT      ADC     F
IT_ADC_F15      IT      ADC     F
IT_ADC_F1_      IT      ADC     F
IT_ADC_F2_      IT      ADC     F
IT_ADC_F3_      IT      ADC     F
IT_ADC_F4_      IT      ADC     F
IT_ADC_F5_      IT      ADC     F
IT_ADC_F6_      IT      ADC     F
IT_ADC_M10      IT      ADC     M
IT_ADC_M11      IT      ADC     M
IT_ADC_M12      IT      ADC     M
IT_ADC_M16      IT      ADC     M
IT_ADC_M17      IT      ADC     M
IT_ADC_M18      IT      ADC     M
IT_ADC_M7_      IT      ADC     M
IT_ADC_M8_      IT      ADC     M
IT_ADC_M9_      IT      ADC     M


```
If you want to rename the filename then you can use mv command. For this use mv file1 file2. This will rename file1 into file2. For e.g. if you do not like the name cols_data.txt, then you can change into beetles.txt by typing mv cols_data.txt beetles.txt

```
[sbhatta4@pbio381 mydata]$ mv cols_data.txt beetles.txt
[sbhatta4@pbio381 mydata]$ beetles.txt


```
If we want to save the output of command grep 'IT' beetles.txt into a separate file, then we should use following commands: grep 'IT' beetles.txt >beetles_ITonly.txt 

```

[sbhatta4@pbio381 mydata]$ grep 'IT' beetles.txt >beetles_ITonly.txt
[sbhatta4@pbio381 mydata]$ ll
total 8
-rw-r--r--. 1 sbhatta4 users  480 Jan 24 14:39 beetles_ITonly.txt
-rwxr-xr-x. 1 sbhatta4 users 1474 Jan 24 14:16 beetles.txt


```

------
<div id='id-section2'/>
### Page 2: 2018-01-29 Starting working with the bull headed beetle. 

### Overview
Raw data of RNA sequencing were provided by Prof. Melissa. The objectives of this 
class were to understand the biology of beetle, to familiarize with the general work flow
or "pipeline" for processing and analyzing RNAseq data and to visualize and interpret Illumina data quality
using a fastq file. 

### The Beetles

Onthophagus taurus beetles are native to the mediterranean and they were introduced to Australia in the late 1960s.
These were dung beetles and helped reduce fly pests and disease that were detrimental to the livestock. In 1970s,
these beetles were "accidentally" introduced to the eastern Unites States from an unknown origin. 

### Study Question
The study aimed to determine the genetic variation among the beetles which were native to Italy (IT), 
Western Australia (WA), and North Carolina (NC). RNA sequencing technique was employed to address this question. 
The samples were sequenced on about 7 lanes of Illumina HiSeq2500. 
### Samples
Three individuals of three given populations (IT, WA, NC) of four developmental stages viz. late third larval instar (L3L), pre-pupae day 1 (PP1), 
pupae day 1 (PD1), and adults four days after ecolsion (AD4) were collected for both the sexes. 
Total number of samples were thus 3*3*4*2 = 72 samples

### Pipeline
This refers to the entire process that the raw sequencing data goes through before we get some meaningful information out of it.
1. Visualizing and Cleaning data.
To visualize the quality of raw data, program FastQC is used and to clean the raw data, program Trimmomatic can be used. Later, the quality of cleaned data can be again visualized in FastQC.

The other steps of Pipeline would be updated after we actually do it.

### Working with RNAseq data
Raw RNA sequencing data from Beeltes were present in the shared space of our pbio381.uvm.edu server. 
One file containing raw RNA sequences were provided to each of us. We used the following codes to locate our files.
```
[sbhatta4@pbio381 data]$ cd project_data/beetles/rawdata/

```
Once we find the file. We used a command to see what's in the file. For this zcat command is used 

```
sbhatta4@pbio381 rawdata]$ zcat WA_PP1_M3_GTGAAA_L003_R2_001.fastq.gz | head -n4
```
In here, zcat is the command, and "WA...gz" is the file name. '| head - n4' means that the first four lines of the file will be shown. 
We found something like this.

```
[sbhatta4@pbio381 rawdata]$ zcat WA_PP1_M3_GTGAAA_L003_R2_001.fastq.gz | head -n4
@CCRI0219:155:C2LNBACXX:3:1101:1412:1969 2:N:0:GTGAAA
GAAAGACCAAAATCCCTTTCGACTGACGGCTTAATGACGGCCCTCAAGGCATCATNAATCATACNGATAGCNGGTNAAAACCACACAATAAAAAAAAACA
+
CCCFFFFFHHHHGJJJJJJJIIIIGIJJIJIJIGJJFIIJIJJJIIIJIIHGHHH#,;;?BCEC#,5?@DC#,<(#,2<@@DDDDDDDDDDDDDDDDDDD

```
Here, the first line starting with '@CCRI.....' is the information about the read.
The second line is the actual DNA sequence that we need to analyze.
The third line always begins with a plus and sometimes may contain the same info as in line 1. In our case it bears only the + sign
The fourth line has the string of characters and each of these character represent the quality scores of the bases that we have in second line.
Note that the characters in fourth line equals the number of bases in second line, and the quality of each base in the second line is given by the character that is present below it in the fourth line. 

To determine the quality of the bases, Phred or Q score are used. Phred score of 10, 20, 30 and 40 represents 90%, 99%, 99.9% and 99.99% base call accuracy respectively. The Phred score of greater than 30 is considered to have a decent quality. Note that 99.99% base call accuracy implies that the machine is 99.99% sure that the base that it called is the same base that's present in the sample. 
The Phred score can is translated to ASCII characters so that the two digit number can be represented by a single character. 

```
Quality encoding: !"#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHI
                  |         |         |         |         |
    Quality score:0........10........20........30........40   
```
We can see that ! refers to the Phred Q score of 0 where as I refers to the score 40. 

### Visualizing the data with FastQC
We have a program FastQC already installed in our /data/popgen directory 
------
<div id='id-section3'/>
### Page 3:
------
<div id='id-section4'/>
### Page 4:
------
<div id='id-section5'/>
### Page 5:
------
<div id='id-section6'/>
### Page 6:
------
<div id='id-section7'/>
### Page 7:
------
<div id='id-section8'/>
### Page 8:
------
<div id='id-section9'/>
### Page 9:
------
<div id='id-section10'/>
### Page 10:
------
<div id='id-section11'/>
### Page 11:
------
<div id='id-section12'/>
### Page 12:
------
<div id='id-section13'/>
### Page 13:
------
<div id='id-section14'/>
### Page 14:
------
<div id='id-section15'/>
### Page 15:
------
<div id='id-section16'/>
### Page 16:
------
<div id='id-section17'/>
### Page 17:
------
<div id='id-section18'/>
### Page 18:
------
<div id='id-section19'/>
### Page 19:
------
<div id='id-section20'/>
### Page 20:
------
<div id='id-section21'/>
### Page 21:
------
<div id='id-section22'/>
### Page 22:
------
<div id='id-section23'/>
### Page 23:
------
<div id='id-section24'/>
### Page 24:
------
<div id='id-section25'/>
### Page 25:
------
<div id='id-section26'/>
### Page 26:
------
<div id='id-section27'/>
### Page 27:
------
<div id='id-section28'/>
### Page 28:
------
<div id='id-section29'/>
### Page 29:
------
<div id='id-section30'/>
### Page 30:
------
<div id='id-section31'/>
### Page 31:
------
<div id='id-section32'/>
### Page 32:
------
<div id='id-section33'/>
### Page 33:
------
<div id='id-section34'/>
### Page 34:
------
<div id='id-section35'/>
### Page 35:
------
<div id='id-section36'/>
### Page 36:
------
<div id='id-section37'/>
### Page 37:
------
<div id='id-section38'/>
### Page 38:
------
<div id='id-section39'/>
### Page 39:
------
<div id='id-section40'/>
### Page 40:
------
<div id='id-section41'/>
### Page 41:
------
<div id='id-section42'/>
### Page 42:
------
<div id='id-section43'/>
### Page 43:
------
<div id='id-section44'/>
### Page 44:
------
<div id='id-section45'/>
### Page 45:
------
<div id='id-section46'/>
### Page 46:
------
<div id='id-section47'/>
### Page 47:
------
<div id='id-section48'/>
### Page 48:
------
<div id='id-section49'/>
### Page 49:
------
<div id='id-section50'/>
### Page 50:
------
<div id='id-section51'/>
### Page 51:
------
<div id='id-section52'/>
### Page 52:
------
<div id='id-section53'/>
### Page 53:
------
<div id='id-section54'/>
### Page 54:
------
<div id='id-section55'/>
### Page 55:
------
<div id='id-section56'/>
### Page 56:
------
<div id='id-section57'/>
### Page 57:
------
<div id='id-section58'/>
### Page 58:
------
<div id='id-section59'/>
### Page 59:
------
<div id='id-section60'/>
### Page 60:

------
